### PR TITLE
Dev/tlvf fix dynamic list

### PR DESCRIPTION
**Pull request by tomereli**
Multiple bug fixes for dynamic lists in tlvf

- Fix potential null pointer dereference when calling class_swap()
when the pointer is uninitialized.
- fix int length lists parse bugs
- add length function for unknown length lists

### CODE REVIEW

File: framework/tlvf/tlvf.py [hunk 1]
```diff
@@ -1007,10 +1007,6 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
             lines_cpp.append( "%sreturn false;" % self.getIndentation(2))
             lines_cpp.append( "%s}" % self.getIndentation(1) )
 
-            lines_cpp.append( "%sif (ptr->getStartBuffPtr() != (uint8_t *)m_%s) {" % (self.getIndentation(1), param_name) )
```
Comment (LiorAmram) [inline framework/tlvf/tlvf.py]: Why did you remove it? Moran's change fixes the bug in this code

Comment (tomereli) [reply framework/tlvf/tlvf.py]: I removed it since it is not correct, parsing fails for list of classes of multiple entries. Once Moran's change is merged, I'll rebase and we'll get it back.

File: framework/tlvf/tlvf.py [hunk 2]
```diff
@@ -851,10 +851,11 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
             if (param_val_const or param_val):
                 if is_int_len or is_const_len:
                     if param_type_info.type != TypeInfo.STD: self.abort("%s.yaml --> only std types are allowed in val / val_const" % self.yaml_fname)
-                    lines_cpp += ["for (size_t i = 0; i < %s; i++){" % (str(param_length)),
-                                 "%sm_%s[i] = %s;"% (self.getIndentation(1), param_name, str(param_val) if param_val != None else str(param_val_const)),
-                                 "}",
-                                 ""]
+                    lines_cpp += ["if (!m_parse__) {",
```
Comment (arnout) [inline framework/tlvf/tlvf.py:854]: It would be clearer both for the generated code and the code generator itself if the parse condition was there only once. So for tlvf, that means to swap the conditions `if (param_val_const or param_val)` and `if is_int_len or is_const_len` and then to merge the two int/const len conditions that follow each other, and then to move the generated `if (!m_parse__)` to the beginning of that condition. Am I making sense?

(nitpick).

File: framework/tlvf/yaml/tlvf/test/tlvVarList.yaml [hunk 3]
```diff
@@ -38,3 +38,7 @@ cInner:
     _type: uint8_t
     _length: [ list_length ]
   var1: uint32_t
+  unknown_length_list:
```
Comment (arnout) [inline framework/tlvf/yaml/tlvf/test/tlvVarList.yaml:41]: Could you give a different name than the unkown_length_list that already exists in the outer class? Otherwise the code gets a little bit confusing.

The same problem already exists for var1, with things like `var1()->var1()` in the code...

File: framework/tlvf/yaml/tlvf/test/tlvVarList.yaml [hunk 4]
```diff
@@ -38,3 +38,7 @@ cInner:
...
+    _type: uint8_t
+    _length: []
```
Comment (arnout) [inline framework/tlvf/yaml/tlvf/test/tlvVarList.yaml:43]: This cannot possibly be parsed. An unknown length list can only exist as the last field in a TLV, because it uses the TLV length to determine the length of the list. Here, cInner is used in e.g. the outer class's var1, so there is no way where the boundary lies between var1's unknown_length_list and the outer unkown_length_list. (Note the confusion with the names :-).

**Review by arnout — CHANGES_REQUESTED** [review]
Commit 1-3 are good to go, except that the wrapping of commit 3's message is not correct.

I would like to see a test for the issues fixed in commit 1-3. Preferably as a new commit that comes before the rest, so we see the test break in the current situation. For the first one, we can just reuse the existing tlvMacAddress.

Commit 4 needs to be fixed.

File: framework/tlvf/tlvf.py (hunk 1 above)
Comment (tomereli) [reply framework/tlvf/tlvf.py]: Brought it back

**Comment by tomereli** [comment]
> Commit 1-3 are good to go, except that the wrapping of commit 3's message is not correct.
> 
> I would like to see a test for the issues fixed in commit 1-3. Preferably as a new commit that comes before the rest, so we see the test break in the current situation. For the first one, we can just reuse the existing tlvMacAddress.
> 
> Commit 4 needs to be fixed.

Commit 1 - no test I can think of since it doesn't really happen, it is just safer to check the pointer we are actually using.
Commit 2 - added a test
Commit 3 - Just adding a length getter function, no test can be added.

File: tools/prplMesh.code-workspace [hunk 5]
```diff
@@ -77,7 +77,10 @@
 			"utility": "cpp",
 			"atomic": "cpp",
 			"codecvt": "cpp",
-			"memory": "cpp"
+			"memory": "cpp",
```
Comment (tomereli) [inline tools/prplMesh.code-workspace:80]: Not related to the PR - Please ignore

**Comment by tomereli** [comment]
> Commit 1-3 are good to go, except that the wrapping of commit 3's message is not correct.
> 
> I would like to see a test for the issues fixed in commit 1-3. Preferably as a new commit that comes before the rest, so we see the test break in the current situation. For the first one, we can just reuse the existing tlvMacAddress.
> 
> Commit 4 needs to be fixed.

I opened a new PR  #144 for commits 1-3 - please review, then i'll rebase and fix this one.

**Comment by tomereli** [comment]
Fixed in #148
